src/Lean/Hygiene.lean

```diff
@@ -33,4 +33,77 @@ instance MonadQuotation : MonadQuotation Unhygienic := {
 protected def run {α : Type} (x : Unhygienic α) : α := run x firstFrontendMacroScope (firstFrontendMacroScope+1)
 end Unhygienic
 
+private def mkInaccessibleUserNameAux (unicode : Bool) (name : Name) (idx : Nat) : Name :=
+if unicode then
+  if idx == 0 then
+    name.appendAfter "✝"
+  else
+    name.appendAfter ("✝" ++ idx.toSuperscriptString)
+else
+  name ++ mkNameNum "_inaccessible" idx
+
+private def mkInaccessibleUserName (unicode : Bool) : Name → Name
+| Name.num p@(Name.str _ _ _) idx _ =>
+  mkInaccessibleUserNameAux unicode p idx
+| Name.num Name.anonymous idx _     =>
+  mkInaccessibleUserNameAux unicode Name.anonymous idx
+| Name.num p idx _ =>
+  if unicode then
+    (mkInaccessibleUserName p).appendAfter ("⁻" ++ idx.toSuperscriptString)
+  else
+    mkNameNum (mkInaccessibleUserName p) idx
+| n => n
+
+@[export lean_is_inaccessible_user_name]
+def isInaccessibleUserName : Name → Bool
+| Name.str _ s _   => s.contains '✝' || s == "_inaccessible"
+| Name.num p idx _ => isInaccessibleUserName p
+| _                => false
+
+def sanitizeNamesDefault := true
+@[init] def sanitizeNamesOption : IO Unit :=
+registerOption `pp.sanitizeNames { defValue := sanitizeNamesDefault, group := "pp", descr := "add suffix '_{<idx>}' to shadowed/inaccessible variables when pretty printing" }
+def getSanitizeNames (o : Options) : Bool:= o.get `pp.sanitizeNames sanitizeNamesDefault
+
+structure NameSanitizerState :=
+(options            : Options)
+-- `x` ~> 2 if we're already using `x✝`, `x✝¹`
+(nameStem2Idx       : NameMap Nat := {})
+-- `x._hyg...` ~> `x✝`
+(userName2Sanitized : NameMap Name := {})
```

Comment:
@leodemoura I bundled up the sanitizer state and use the StateM monad throughout so we can compose multiple sanitization steps like Syntax + LocalContext, even if I [didn't find](https://github.com/leanprover/lean4/issues/182#issuecomment-696680071) any of these yet.

Reply:
The `formatAux` function must track this state. At the `withContext` case we sanitize a `LocalContext`.
```lean
| nCtx, _,         withContext ctx d        => formatAux nCtx (some $ sanitizeNames ctx) d
```

Reply:
```lean
partial def formatAux : NamingContext → Option MessageDataContext → MessageData → StateRefT NameSanitizerState IO Format
```

Reply:
Yes, this is the plan. I'm still working on coming up with a good example.

Reply:
I was trying to find a good example too and found a bug at `pp.macroStack`. I will push a fix.

Reply:
Did you find an example?

Reply:
Not sure, it is hard to tell on my branch since the scopes are erased when I pretty-print the macro stack. The test `macroStack.lean` is the one I used. I will try again with this PR.